runtime/src/pinned_threads.rs

```diff
@@ -0,0 +1,74 @@
+use rayon::ThreadPool;
+use rayon_core::{ThreadBuilder, ThreadPoolBuilder};
+use std::{io, thread};
+
+const NUM_THREADS_PER_CORE: usize = 8;
+const MAX_NUM_OF_THREADS: usize = 128;
+
+#[derive(Debug, Default)]
+pub struct PinnedSpawn {
+    cores: Vec<core_affinity::CoreId>,
+    len: usize,
+    core_id_pointer: usize,
+}
+
+impl PinnedSpawn {
+    // pub fn new(num_cores: usize) -> Self {
+    //     let core_ids = core_affinity::get_core_ids().unwrap();
+    //     if num_cores > core_ids.len() {
+    //         panic!("More cores requested than available");
+    //     }
+    //     Self {
+    //         cores: core_ids.into_iter().rev().take(num_cores).collect(),
+    //         len: num_cores,
+    //         core_id_pointer: 0,
+    //     }
+    // }
+
+    // Pins as many threads as the ceil of the fraction times the total number of cores
+    // This ensures that at least 1 core would be pinned
+    pub fn new_frac_of_cores(num: usize, denom: usize) -> Self {
+        if num > denom {
+            panic!("fraction must be <= 1");
+        }
+        let core_ids = core_affinity::get_core_ids().unwrap();
+        let num_cores = (num * core_ids.len() - 1) / denom + 1;
+        Self {
+            cores: core_ids.into_iter().rev().take(num_cores).collect(),
+            len: num_cores,
+            core_id_pointer: 0,
+        }
+    }
+
+    // Spawn threads pinned to core in a round robin fashion
+    pub fn spawn(&mut self, thread: ThreadBuilder) -> io::Result<()> {
+        let mut b = thread::Builder::new();
+        if let Some(name) = thread.name() {
+            b = b.name(name.to_owned());
+        }
+        if let Some(stack_size) = thread.stack_size() {
+            b = b.stack_size(stack_size);
+        }
+        let id_for_spawn = self.cores[self.core_id_pointer];
+        b.spawn(move || {
+            core_affinity::set_for_current(id_for_spawn);
```

Comment:
Could you expand on in a comment what the problem we're trying to solve with pinning the threads/benefit of pinning these threads to particular cores would be? Also perhaps a new bench in `runtime/benches/accounts.rs` to demonstrate this difference?

> Check no perf regression when running bench TPS
Might also be a good idea to try running this on a full validator to see if there are any unexpected drawdowns in other components of the validator.

Reply:
Hey @carllin 

My apologies that the thoughts are a little jumbled and still evolving on the issue. Will eventually summarise investigations in a report of sorts - there is a larger question of avoiding mmap, as do most modern SSD-oriented DBs. And it ought to at least be spelled out what the potential benefits are. The biggest benefit may be to reduce memory requirement while improving the worst-case critical path (this is related to some security, performance, and pricing issues, such as the tx cost model).

---

Pinning the threads in particular was just an idea to reduce resource contention, so that the loading threads don’t eat up the resources of the entire CPU. 

However, I’ve found by measuring the time to perform `load_accounts_parallel` when running bench-tps that pinning more than 2 threads per core makes things worse. But, pinning 2 threads to each CPU rather than to a subset of the CPUs is actually the best option for the thread pool (250-500us for 128 tx batch). One can avoid core 0 for PoH.

Note that in the replay stage, a different scenario arises. One can no longer “cheat” by parallelising across threads. That’s because the default behaviour is to already saturate the threads with `load_and_execute`. What this means is that essentially, every job

Results (approximate from eyeballing):
| Description | Time for 128 tx `load_accounts_parallel` (banking stage) | (replay stage) |
|---|---|---|
| Pin 2 threads to all logical cores |250-500us| avg: 12_000us tail: 25_000us|
| Pin 1 thread to all logical cores |250-500us| avg: 25_000us tail: 50_000us|
| Pin 4 threads to all logical cores | 800-1800us | avg: 25_000us tail 50_000us|
| Pin 1 thread to 1/4 of logical cores | 800-1800us | NIL |
| Ordinary thread pool 1 thread per logical core | 800-1500us | avg: 25_000us tail 50_000us|
| Ordinary thread pool 2 threads per logical core | 800-1500us | avg: 12_000us tail 25_000us|
| Single thread | 1800-5000us | 1800-5000us |

Reply:
As it is difficult to manually trigger page faults, I decided to simulate a single 250us page faults on every account load. 
```rust
pub fn get_account<'a>(&'a self, offset: usize) -> Option<(StoredAccountMeta<'a>, usize)> {
        std::thread::sleep(std::time::Duration::from_micros(250));
```
Here are my results (32 IO threads on 8C16T machine):
This PR: 10_000-25_000us (banking). Master 350_000-500_000us. 

I honestly don't have an explanation for why a factor of 40 increase in page loading is observed...